### PR TITLE
Simplify our buildSrc

### PR DESCRIPTION
**Pull request by BraisGabin**
The idea is to simplify how we configure our own gradle.

Right now we force that all our modules have a specific configuration and with this new approuch we provide a plugin that configures the module as we want and is the module that opts-in this configuration.

An advantage is that we don't need to add exceptions for `:detekt-bom`. It just don't use this module.

The idea is that this is just the first step. Once we do this I'll like to use the plugin https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin to improve how we manage our dependencies because we add a lot by default and we probably don't need that much. And, probably, we could remove all the code in `commons.gradle.kts` at some point.

This will help with the integration with the plugin `binary-compatibility-validator` too. This last one was the trigger to start this refactor.

### CODE REVIEW

File: buildSrc/src/main/kotlin/module.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    archives(sourcesJar)
+    archives(javadocJar)
+}
+
+publishing {
+    publications.named<MavenPublication>(DETEKT_PUBLICATION) {
+        from(components["java"])
+        artifact(sourcesJar.get())
+        artifact(javadocJar.get())
+        if (project.name == "detekt-cli") {
```
Comment (chao2zhang) [inline buildSrc/src/main/kotlin/module.gradle.kts:100]: Could we refactor this condition to `detekt-cli`?

File: buildSrc/src/main/kotlin/packaging.gradle.kts [hunk 2]
```diff
@@ -95,32 +82,3 @@ subprojects {
         logger.info("Signing Disabled as the PGP key was not found")
     }
 }
-
-configure(subprojects.filter { it.name != "detekt-bom" }) {
```
Comment (chao2zhang) [inline buildSrc/src/main/kotlin/packaging.gradle.kts:99]: This looks great!

File: buildSrc/src/main/kotlin/module.gradle.kts (hunk 1 above)
Comment (BraisGabin) [reply buildSrc/src/main/kotlin/module.gradle.kts:100]: Yes, I agree that this should be removed too. But I'd prefer to do it in other PR.

**Review by schalkms — APPROVED** [review]
Great stuff 👏

File: buildSrc/src/main/kotlin/module.gradle.kts (hunk 1 above)
Comment (BraisGabin) [reply buildSrc/src/main/kotlin/module.gradle.kts:99]: Maybe we can remove this if directly: #3329

**Comment by chao2zhang** [comment]
What's the recommended merging order between this PR and https://github.com/detekt/detekt/pull/3320? I am open to merge in this PR first then see if the hack in  #3320 would be fixed

**Comment by BraisGabin** [comment]
I think that this is a big change and we should wait for other maintainer to approve it. I know that @arturbosch worked a lot refactoring our gradle scripts (they were a real mess all in one file) so I think that we should wait a bit to merge this.

**Comment by chao2zhang** [comment]
> I think that this is a big change and we should wait for other maintainer to approve it. I know that @arturbosch worked a lot refactoring our gradle scripts (they were a real mess all in one file) so I think that we should wait a bit to merge this.

Is it a good time to merge #3320 consequently? It creates merge conflicts for a few PRs already.

File: build.gradle.kts [hunk 3]
```diff
@@ -4,11 +4,15 @@ plugins {
     releasing
     detekt
     id("org.jetbrains.dokka") apply false
-    id("com.github.johnrengelman.shadow") apply false
+    id("com.github.johnrengelman.shadow") version "5.2.0" apply false
```
Comment (arturbosch) [inline build.gradle.kts:7]: This introduces a new place to remember where a version needs to be changed.
Is this change necessary?

**Review by arturbosch — APPROVED** [review]
Looks good!

File: build.gradle.kts (hunk 3 above)
Comment (BraisGabin) [reply build.gradle.kts]: I don't remember why I did that. Probably because we don't need the shadow pugin in buildSrc now. But I moved it to the buildSrc again 👍
